### PR TITLE
add contract author and version flags

### PR DESCRIPTION
**Pull request by salaheldinsoliman**
This PR implements a way to specify the author and version of all contracts via the CLI or via "solang.toml"

### CODE REVIEW

File: docs/running.rst [hunk 1]
```diff
@@ -154,6 +154,16 @@ Options:
...
 
+\-\-contract-authors
+   Specify authors for all contracts. If a `@author` tag is present, it will override this argument for the targeted contract.
+
+  .. note::
+
+      This will only affect the metadata in case of substrate target.
+
+\-\-version
+   Specify contracts version.
```
Comment (LucasSte) [inline docs/running.rst:165]:
```suggestion
   Specify contracts' version.
```

**Review by LucasSte — APPROVED** [review]

File: docs/running.rst [hunk 2]
```diff
@@ -154,6 +154,16 @@ Options:
 \-\-wasm-opt
    wasm-opt passes for Wasm targets (0, 1, 2, 3, 4, s or z; see the wasm-opt help for more details).
 
+\-\-contract-authors
+   Specify authors for all contracts. If a `@author` tag is present, it will override this argument for the targeted contract.
+
+  .. note::
+
+      This will only affect the metadata in case of substrate target.
```
Comment (seanyoung) [inline docs/running.rst:162]: Then the command line option should not be allowed

**Review by seanyoung — CHANGES_REQUESTED** [review]
This is broken.

```
$ cargo run compile --target substrate --contract-authors foo,bar examples/substrate/flipper.sol 
    Finished dev [unoptimized + debuginfo] target(s) in 0.07s
     Running `target/debug/solang compile --target substrate --contract-authors foo,bar examples/substrate/flipper.sol`
No input files specified, please specifiy them in solang.toml or in command line

File: src/bin/solang.rs [hunk 3]
```diff
@@ -241,6 +242,21 @@ fn compile(compile_args: &Compile) {
...
+        let authors = if let Some(authors) = &compile_args.package.authors {
+            if target == Target::Solana {
+                eprintln!("warning: the `authors` flag will be ignored for solana target")
+            }
+            authors.clone()
+        } else {
+            vec!["unknown".to_string()]
+        };
+
+        let version = if let Some(version) = &compile_args.package.version {
```
Comment (xermicus) [inline src/bin/solang.rs:254]: Nit, can use `unwrap_or`?

File: docs/running.rst [hunk 4]
```diff
@@ -154,6 +154,16 @@ Options:
...
+
+\-\-version
```
Comment (xermicus) [inline docs/running.rst:164]: This should mention that a specific format is expected (also which one).

File: src/lib.rs [hunk 5]
```diff
@@ -158,7 +158,14 @@ pub fn compile(
         if contract.instantiable {
             let code = contract.emit(&ns, &opts);
 
-            let (abistr, _) = abi::generate_abi(contract_no, &ns, &code, false);
+            let (abistr, _) = abi::generate_abi(
+                contract_no,
+                &ns,
+                &code,
+                false,
+                &vec!["unknown".to_string()],
```
Comment (xermicus) [inline src/lib.rs:166]: Could use the default values here. Also it'd be nice if this would work in the library version too.

**Review by xermicus — COMMENTED** [review]
LGTM

File: src/bin/solang.rs (hunk 3 above)
Comment (salaheldinsoliman) [reply src/bin/solang.rs:254]: It looks better, but it wants to take ownership of `compile_args.package.version`. Then inside `unwrap_or`, it needs to return a reference, so I have to use a let statement beforehand. E.G the final product would look like this: 
`let default_version = "0.0.1".to_string();
 let version = compile_args.package.version.as_ref().unwrap_or(&default_version);`
 So I think `if let some` works better here

Comment (xermicus) [reply src/bin/solang.rs:254]: Not terribly important, the `if let` is fine too

**Review by seanyoung — APPROVED** [review]

**Review by xermicus — APPROVED** [review]
Please include what versioning schema is expected in the docs

**Pull request merged**